File: observer/golang/pkg/observer/observer.go
package observer

import (
	stderr "errors"
	"github.com/veresnikov/ood-labs/observer/golang/pkg/utils"
	"golang.org/x/exp/maps"
	"golang.org/x/exp/slices"
)

var (
	ErrObserverRegistered = stderr.New("observer already registered")
	ErrObserverNotFound   = stderr.New("observer not found")
)

type Observer[T any] interface {
	Update(data T)
}

type Observable[T any] interface {
	RegisterObserver(observer Observer[T], priority int) error
	RemoveObserver(observer Observer[T]) error
	NotifyObservers()
}

func NewBaseObservable[T any](getChangeData func() T) Observable[T] {
	return &baseObservable[T]{
		observers:      make(map[int][]Observer[T]),
		getChangedData: getChangeData,
	}
}

type baseObservable[T any] struct {
	observers           map[int][]Observer[T]
	availablePriorities []int
	getChangedData      func() T
}

func (base *baseObservable[T]) RegisterObserver(observer Observer[T], priority int) error {
	for _, observers := range base.observers {
		_, err := utils.Find(observers, observer)
		if err == nil {
			return ErrObserverRegistered
		}
	}
	base.observers[priority] = append(base.observers[priority], observer)
	base.updateAvailablePriorities()
	return nil
}

func (base *baseObservable[T]) RemoveObserver(observer Observer[T]) error {
	for priority, observers := range base.observers {
		updatedObservers, err := utils.Remove(observers, observer)
		if err == nil {
			base.observers[priority] = updatedObservers

			if len(base.observers[priority]) == 0 {
				delete(base.observers, priority)
			}
			base.updateAvailablePriorities()
			return nil
		}
	}
	return ErrObserverNotFound
}

func (base *baseObservable[T]) NotifyObservers() {
	data := base.getChangedData()
	availableObservers := base.observers
	for _, key := range base.availablePriorities {
		for _, observer := range availableObservers[key] {
			observer.Update(data)
		}
	}
}

func (base *baseObservable[T]) updateAvailablePriorities() {
	base.availablePriorities = maps.Keys(base.observers)
	slices.Sort(base.availablePriorities)
}

File: mvc/src/common/observer/observer.ts
abstract class Observable<Type> {
    private observers: Observer<Type>[] = []

    RegisterObserver(observer: Observer<Type>): void {
        let index = this.observers.findIndex((item) => {
            return item === observer
        })
        if (index === -1) {
            this.observers.push(observer)
            this.InitializeObserver(observer)
            return
        }
        throw new Error("observer already registered")
    }

    RemoveObserver(observer: Observer<Type>): void {
        let index = this.observers.findIndex((item) => {
            return item === observer
        })
        if (index === -1) {
            throw new Error("observer not found")
        }
        this.observers.splice(index, 1)
    }

    NotifyObserver(): void {
        const data = this.GetChangedData()
        this.observers.forEach((observer) => {
            observer.Update(data)
        })
    }

    private InitializeObserver(observer: Observer<Type>): void {
        const data = this.GetChangedData()
        observer.Update(data)
    }

    protected abstract GetChangedData(): Type;
}

abstract class Observer<Type> {
    abstract Update(data: Type): void;
}

export {Observable, Observer}